sled-agent/src/services.rs

```diff
@@ -3682,7 +3723,7 @@ mod test {
     const GLOBAL_ZONE_BOOTSTRAP_IP: Ipv6Addr = Ipv6Addr::LOCALHOST;
     const SWITCH_ZONE_BOOTSTRAP_IP: Ipv6Addr = Ipv6Addr::LOCALHOST;
 
-    const EXPECTED_ZONE_NAME_PREFIX: &str = "oxz_oximeter";
+    const EXPECTED_ZONE_NAME_PREFIX: &str = "oxz_ntp";
```

Comment:
Apparently the mock tests don't work very well with self-assembled zones.

I attempted to adapt the tests to work with self assembled zones but it appears to be that this is a much larger body of work out of the scope of this PR. Additionally, we can confirm this new zone works as the "helios / deploy" test is passing.

This is what the tests fail with:


> Could not create service: ZoneInitialize([("oxz_oximeter_181462b0-2f83-4cc8-9c95-95884a038c26", Io { message: "Failed to setup Oximeter profile", err: Os { code: 2, kind: NotFound, message: "No such file or directory" } })])

Output from the logs:

> {"msg":"Creating ServiceManager","v":0,"name":"test_ensure_service","level":30,"time":"2024-01-17T22:13:09.828335743Z","hostname":"pop-os","pid":676988,"component":"ServiceManager"}
{"msg":"sled agent started","v":0,"name":"test_ensure_service","level":30,"time":"2024-01-17T22:13:09.828532057Z","hostname":"pop-os","pid":676988,"component":"ServiceManager","underlay_address":"::1"}
{"msg":"new DNS resolver","v":0,"name":"test_ensure_service","level":30,"time":"2024-01-17T22:13:09.828553155Z","hostname":"pop-os","pid":676988,"component":"DnsResolver","component":"ServiceManager","addresses":"[[::1:0:0:0:1]:53, [::2:0:0:0:1]:53, [::3:0:0:0:1]:53, [::4:0:0:0:1]:53, [::5:0:0:0:1]:53]"}
{"msg":"No ledger in /tmp/.tmpvxo60Q/omicron-zones.json","v":0,"name":"test_ensure_service","level":30,"time":"2024-01-17T22:13:09.828568418Z","hostname":"pop-os","pid":676988,"component":"ServiceManager"}
{"msg":"Failed to read ledger: Not found in storage","v":0,"name":"test_ensure_service","level":20,"time":"2024-01-17T22:13:09.828581612Z","hostname":"pop-os","pid":676988,"component":"ServiceManager","path":"/tmp/.tmpvxo60Q/omicron-zones.json"}
{"msg":"No ledger in /tmp/.tmpvxo60Q/omicron-zones.json","v":0,"name":"test_ensure_service","level":30,"time":"2024-01-17T22:13:09.828601728Z","hostname":"pop-os","pid":676988,"component":"ServiceManager"}
{"msg":"Failed to read ledger: Not found in storage","v":0,"name":"test_ensure_service","level":20,"time":"2024-01-17T22:13:09.828614765Z","hostname":"pop-os","pid":676988,"component":"ServiceManager","path":"/tmp/.tmpvxo60Q/omicron-zones.json"}
{"msg":"ensure_all_omicron_zones_persistent","v":0,"name":"test_ensure_service","level":20,"time":"2024-01-17T22:13:09.828633403Z","hostname":"pop-os","pid":676988,"component":"ServiceManager","ledger_generation":"1","request_generation":"2"}
{"msg":"Configured to skip timesync checks","v":0,"name":"test_ensure_service","level":30,"time":"2024-01-17T22:13:09.828649126Z","hostname":"pop-os","pid":676988,"component":"ServiceManager"}
{"msg":"top of bundle cleanup loop","v":0,"name":"test_ensure_service","level":30,"time":"2024-01-17T22:13:09.828823024Z","hostname":"pop-os","pid":676988,"component":"auto-cleanup-task","time_to_next_cleanup":"599.999378173s","next_cleanup":"Instant { tv_sec: 263767, tv_nsec: 242078616 }"}
{"msg":"Profile for oxz_oximeter_181462b0-2f83-4cc8-9c95-95884a038c26:\n<!DOCTYPE service_bundle SYSTEM \"/usr/share/lib/xml/dtd/service_bundle.dtd.1\">\n<service_bundle type=\"profile\" name=\"omicron\">\n  <service version=\"1\" type=\"service\" name=\"oxide/zone-network-setup\">\n      <property_group type=\"application\" name=\"config\">\n        <propval type=\"astring\" name=\"datalink\" value=\"oxControlService0\"/>\n        <propval type=\"astring\" name=\"gateway\" value=\"::1\"/>\n        <propval type=\"astring\" name=\"static_addr\" value=\"::1\"/>\n      </property_group>\n    <instance enabled=\"true\" name=\"default\">\n    </instance>\n  </service>\n  <service version=\"1\" type=\"service\" name=\"network/ssh\">\n    <instance enabled=\"false\" name=\"default\">\n    </instance>\n  </service>\n  <service version=\"1\" type=\"service\" name=\"oxide/oximeter\">\n    <instance enabled=\"true\" name=\"default\">\n      <property_group type=\"application\" name=\"config\">\n        <propval type=\"astring\" name=\"id\" value=\"181462b0-2f83-4cc8-9c95-95884a038c26\"/>\n        <propval type=\"astring\" name=\"address\" value=\"[::1]:12223\"/>\n      </property_group>\n    </instance>\n  </service>\n</service_bundle>","v":0,"name":"test_ensure_service","level":30,"time":"2024-01-17T22:13:09.829194091Z","hostname":"pop-os","pid":676988,"component":"ServiceManager"}

Reply:
Yeah, I've traditionally used this function to manipulate the local filesystem paths to make any local file checking happier:

https://github.com/oxidecomputer/omicron/blob/213d5cc33f8311a0033212777da9686a9469a3c2/sled-agent/src/services.rs#L3944-L3955

Admittedly, this is trickier to do when manipulating paths in zones that haven't been started.  I *suspect* that's what you're hitting -- instead of putting the profile xml file into the zone, we're finding that some `/zone` path doesn't exist.

I think "swapping the zone we're acting on" is fine, but if we make NTP self-assembling, we'll probably bump into this issue again.

I'd be fine with either:
- Injecting a hook to manipulate "where that file gets dropped", so the test can inspect it, or
- Doing a bigger overhaul of the tests, to validate the properties we want to check

Reply:
> I think "swapping the zone we're acting on" is fine, but if we make NTP self-assembling, we'll probably bump into this issue again.

yep! my idea was to do:

> Doing a bigger overhaul of the tests, to validate the properties we want to check

In a follow up PR dedicated only to tests before porting the NTP zone to be self-assembling.